File: pathlab/algorithms/GreedyBFS.js
/**
 * PathLab — Greedy Best-First Search
 *
 * Like A* but sorts the open set by h(n) alone — the estimated
 * cost to the goal — ignoring the cost already incurred g(n).
 *
 * This makes it very fast at reaching the goal in simple environments,
 * but sacrifices optimality. It can be misled by the heuristic and
 * produce paths that are significantly longer than optimal.
 *
 * Time:  O(b^m) worst case, often much better in practice
 * Space: O(b^m)
 * Optimal: No — may find a suboptimal path
 * Complete: Yes (finite graph without cycles causing infinite loops)
 */
import { AlgorithmBase } from "./AlgorithmBase.js";
import { MinHeap } from "../core/MinHeap.js";
import { Grid } from "../core/Grid.js";

export class GreedyBFS extends AlgorithmBase {
  *solve() {
    const { grid, dirs8, h } = this;
    const start = this.startCell;
    const goal = this.goalCell;

    const trackPeak = () => {
      // Use the specific variables relevant to the algorithm (e.g., openSet, visited, gScore)
      peakEntries = Math.max(
        peakEntries,
        openSet.size ?? 0,
        visited?.size ?? 0,
      );
    };

    if (!start || !goal) {
      yield {
        type: "done",
        success: false,
        path: [],
        nodesExpanded: 0,
        pathCost: this.pathCost(path) ?? Infinity,
        peakMemoryBytes: this.estimateBytes(peakEntries),
      };
      return;
    }

    let peakEntries = 0;

    const startKey = Grid.key(start);

    // Min-heap ordered solely by heuristic h(n)
    const openSet = new MinHeap((a, b) => a.hVal - b.hVal);
    const visited = new Set([startKey]);
    const cameFrom = new Map([[startKey, null]]);

    openSet.push({ hVal: h(start, goal), cell: start });
    trackPeak();

    let nodesExpanded = 0;

    while (!openSet.empty) {
      const { cell: current } = openSet.pop();
      nodesExpanded++;

      // Goal check
      if (this.isGoal(current)) {
        const path = this.reconstructPath(cameFrom, current);
        for (let i = 1; i < path.length; i++) {
          yield { type: "path", r: path[i].r, c: path[i].c };
        }
        yield {
          type: "done",
          success: true,
          path,
          nodesExpanded,
          pathCost: this.pathCost(path),
          peakMemoryBytes: this.estimateBytes(peakEntries),
        };
        return;
      }

      yield {
        type: "visit",
        r: current.r,
        c: current.c,
        cost: h(current, goal),
      };

      for (const nb of grid.neighbors(current.r, current.c, dirs8)) {
        const key = Grid.key(nb);
        if (!visited.has(key)) {
          visited.add(key);
          cameFrom.set(key, current);
          peakEntries = Math.max(peakEntries, visited.size, openSet.size);
          openSet.push({ hVal: h(nb, goal), cell: nb });
          trackPeak();
        }
      }
    }

    yield {
      type: "done",
      success: false,
      path: [],
      nodesExpanded,
      pathCost: this.pathCost(path) ?? Infinity,
      peakMemoryBytes: this.estimateBytes(peakEntries),
    };
  }
}

File: pathlab/python/greedy_bfs.py
"""
PathLab — Greedy Best-First Search
Python reference implementation matching the JS version.
"""
import heapq
import math

DIRS_4 = [(-1, 0), (1, 0), (0, -1), (0, 1)]
DIRS_8 = [(-1,-1),(-1, 0),(-1, 1),
          ( 0,-1),         ( 0, 1),
          ( 1,-1),( 1, 0),( 1, 1)]


def manhattan(a, b)  -> float: return abs(a[0]-b[0]) + abs(a[1]-b[1])
def euclidean(a, b)  -> float: return math.sqrt((a[0]-b[0])**2+(a[1]-b[1])**2)
def chebyshev(a, b)  -> float: return max(abs(a[0]-b[0]), abs(a[1]-b[1]))

HEURISTICS = {"manhattan": manhattan, "euclidean": euclidean, "chebyshev": chebyshev}


def edge_cost(r0, c0, r1, c1) -> float:
    return math.sqrt(2) if abs(r1 - r0) + abs(c1 - c0) > 1 else 1.0


def greedy_bfs(grid: list[list[int]],
               start: tuple[int, int],
               goal:  tuple[int, int],
               heuristic: str = "manhattan",
               dirs8:     bool = False) -> tuple[list | None, dict]:
    """
    Greedy Best-First Search.

    Sorts open set by h(n) only — does NOT track g(n).
    Fast but not guaranteed to find the optimal path.

    Args:
        grid:      2-D list (0 = open, 1 = wall).
        start:     (row, col).
        goal:      (row, col).
        heuristic: 'manhattan' | 'euclidean' | 'chebyshev'.
        dirs8:     8-directional movement if True.

    Returns:
        (path, stats)
    """
    rows, cols = len(grid), len(grid[0])
    moves = DIRS_8 if dirs8 else DIRS_4
    h     = HEURISTICS.get(heuristic, manhattan)

    visited   = {start}
    came_from = {start: None}

    # (h_val, node)
    pq = [(h(start, goal), start)]
    nodes_expanded = 0

    while pq:
        _, current = heapq.heappop(pq)
        nodes_expanded += 1

        if current == goal:
            path = _reconstruct(came_from, goal)
            cost = _path_cost(path, dirs8)
            return path, {
                "nodes_expanded": nodes_expanded,
                "path_cost":  cost,
                "path_length": len(path),
            }

        r, c = current
        for dr, dc in moves:
            nr, nc = r + dr, c + dc
            nb = (nr, nc)
            if not (0 <= nr < rows and 0 <= nc < cols): continue
            if grid[nr][nc] == 1: continue
            if nb in visited: continue

            visited.add(nb)
            came_from[nb] = current
            heapq.heappush(pq, (h(nb, goal), nb))

    return None, {
        "nodes_expanded": nodes_expanded,
        "path_cost": math.inf,
        "path_length": 0,
    }


def _reconstruct(came_from, current):
    path = []
    while current is not None:
        path.append(current)
        current = came_from[current]
    return path[::-1]


def _path_cost(path, diagonal=False) -> float:
    cost = 0.0
    for i in range(1, len(path)):
        r0, c0 = path[i-1]
        r1, c1 = path[i]
        cost += math.sqrt(2) if abs(r1-r0)+abs(c1-c0) > 1 else 1.0
    return cost


if __name__ == "__main__":
    G = [
        [0, 0, 0, 0, 0, 0, 0],
        [0, 1, 1, 1, 0, 1, 0],
        [0, 0, 0, 1, 0, 1, 0],
        [0, 1, 0, 0, 0, 0, 0],
        [0, 0, 0, 1, 1, 1, 0],
    ]
    path, stats = greedy_bfs(G, (0, 0), (4, 6), heuristic="manhattan")
    print("Greedy BFS path:", path)
    print("Stats:          ", stats)
    print("(Note: path may not be optimal)")
